### PR TITLE
[BEAM-1958] Standard IO Metrics in Java SDK

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by aviemzur** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
R: @bjchambers

**Comment by aviemzur** [comment]
FYI: I added `getName()` to `Metric` to enable users to get the names of the metrics for querying as suggested by @dhalperi. To support this I had to delete some unused code in the `*Cell` classes.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/CounterCell.java [hunk 1]
```diff
@@ -57,28 +57,11 @@ public Long getCumulative() {
...
     add(1);
   }
 
-  @Override
   public void inc(long n) {
     add(n);
   }
-
-  @Override
-  public void dec() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/CounterCell.java]: So counters can never be decremented? Doesn't seem unreasonable, and may make it easier for backends to aggregate, but this seems like an orthogonal PR.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java [hunk 2]
```diff
@@ -236,8 +237,8 @@ public void populateDisplayData(Builder builder) {
     private transient PubsubClient pubsubClient;
 
     private final Counter batchCounter = Metrics.counter(WriterFn.class, "batches");
-    private final Counter elementCounter = Metrics.counter(WriterFn.class, "elements");
-    private final Counter byteCounter = Metrics.counter(WriterFn.class, "bytes");
+    private final Counter elementCounter = IOMetrics.elementsWritten();
```
Comment (bjchambers) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java]: It would be good to clarify that these are elements/bytes written to pubsub, rather than elements/bytes in an input or output `PCollection`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      Metrics.counter(IO_NAMESPACE, BYTES_WRITTEN);
+  private static final Gauge BACKLOG_BYTES_GAUGE = Metrics.gauge(IO_NAMESPACE, BACKLOG_BYTES);
+  private static final Gauge BACKLOG_ELEMENTS_GAUGE = Metrics.gauge(IO_NAMESPACE, BACKLOG_ELEMENTS);
+
+  // ------------------------- Source Metrics -------------------------
+
+  /**
+   * Counter of elements read by a source.
+   */
+  public static Counter elementsRead() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: Maybe call this sourceElementsRead() or something? I'd like to distinguish this from other cases of reading elements:

1. Reading elements from your input collection
2. Reading information from other locations as part of a ParDo that "augments" each element that passes through.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Counter elementsReadBySplit(String splitId) {
+    return Metrics.counter(IO_SPLITS_NAMESPACE, renderName(splitId, ELEMENTS_READ));
+  }
+
+  /**
+   * Counter of bytes read by a source.
+   */
+  public static Counter bytesRead() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: sourceBytesRead for same reason?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Counter bytesReadBySplit(String splitId) {
+    return Metrics.counter(IO_SPLITS_NAMESPACE, renderName(splitId, BYTES_READ));
+  }
+
+  /**
+   * Gauge for source backlog in bytes.
+   */
+  public static Gauge backlogBytes() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: sourceBacklogBytes?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Gauge backlogBytesOfSplit(String splitId) {
+    return Metrics.gauge(IO_SPLITS_NAMESPACE, renderName(splitId, BACKLOG_BYTES));
+  }
+
+  /**
+   * Gauge for source backlog in elements.
+   */
+  public static Gauge backlogElements() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: sourceBacklogElements?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  public static Gauge backlogElementsOfSplit(String splitId) {
+    return Metrics.gauge(IO_SPLITS_NAMESPACE, renderName(splitId, BACKLOG_ELEMENTS));
+  }
+
+  // ------------------------- Sink Metrics -------------------------
+
+  /**
+   * Counter of elements written to a sink.
+   */
+  public static Counter elementsWritten() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: sinkElementsWritten

to distinguish it from "elements output to pcollection"

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    return ELEMENTS_WRITTEN_COUNTER;
+  }
+
+  /**
+   * Counter of bytes written to a sink.
+   */
+  public static Counter bytesWritten() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: sinkBytesWritten

to distinguish it from "bytes written to PCollection" or "bytes written to State"

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/CounterCell.java (hunk 1 above)
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/CounterCell.java]: They can be decremented. This is `CounterCell`, the way `Counter` interface was implemented (`DelegatingCounter`) was using this `CounterCell#inc` with positive/negative values for inc/dec respectively, `CounterCell#dec` was not used.
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metrics.java#L78-L78

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java (hunk 2 above)
Comment (aviemzur) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java]: The metrics are in the `io` namespace and the step in which you write to PubSub. Is that indicative enough?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java (hunk 3 above)
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: Do you mean the name of this factory method, or the name of the metric itself?
For the factory method, is `IOMetrics#elementsRead` not clear? We could perhaps have `SourceMetrics.elementsRead` and `SinkMetrics.elementsWritten` if separating these makes things more coherent.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java (hunk 2 above)
Comment (bjchambers) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java]: Maybe? To make sure I understand, what are your thoughts around a metric such as "number of elements/bytes written to an output PCollection" (or similarly, read from an input PCollection)? Arguably, they may be implemented by the runner and use a different approach than Metrics, so maybe these are clear? I just want to make sure we won't run into conflicts there.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java (hunk 3 above)
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java]: Splitting the factory seems reasonable -- then we also clearly distinguish the source and sink cases from each other, as well as from the case of reading/writing PCollections.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java (hunk 2 above)
Comment (aviemzur) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java]: I wasn't aware these metrics were also on the table.
So we should probably have better namespaces to differentiate the metrics you mentioned from IO metrics.
If the term `io` as a namespace may cause confusion, perhaps we can use `source` and `sink` as namespaces for the IO metrics instead.

**Comment by aviemzur** [comment]
@bjchambers split the factory into 2 classes and renamed namespaces as discussed.
PTAL

**Comment by aviemzur** [comment]
retest this please

**Comment by aviemzur** [comment]
Retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DistributionCell.java [hunk 9]
```diff
@@ -30,11 +30,11 @@
  * of indirection.
  */
 @Experimental(Kind.METRICS)
-public class DistributionCell implements MetricCell<Distribution, DistributionData>, Distribution {
+public class DistributionCell implements MetricCell<DistributionData> {
 
   private final DirtyState dirty = new DirtyState();
   private final AtomicReference<DistributionData> value =
-      new AtomicReference<DistributionData>(DistributionData.EMPTY);
+      new AtomicReference<>(DistributionData.EMPTY);
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DistributionCell.java]: Does this fit on one line now?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metric.java [hunk 10]
```diff
@@ -21,4 +21,7 @@
 /**
  * Marker interface for all user-facing metrics.
  */
-public interface Metric { }
+public interface Metric {
+
+  MetricName getName();
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metric.java]: Can you add Javadoc to this public method?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+  /**
+   * Gauge for source backlog in elements.
+   */
+  public static Gauge backlogElements() {
+    return BACKLOG_ELEMENTS_GAUGE;
+  }
+
+  /**
+   * Gauge for source split backlog in elements.
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java]: Maybe "Should only be used when there is a small, fixed set of split IDs so as not to overload metrics backends."?

(Here and above)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DistributionCell.java (hunk 9 above)
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DistributionCell.java]: No, still too long.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metric.java (hunk 10 above)
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/Metric.java]: 👍

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java (hunk 11 above)
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java]: 👍

**Comment by aviemzur** [comment]
Rebased and made the Javadoc changes requested.
Will merge when Jenkins is alive again and tests pass.

**Comment by aviemzur** [comment]
Retest this please.
